File: apps/server-nest/src/api/model/member-model.ts
import { Member } from '../../domain';
import {
  link,
  Link,
  userHref,
  workspaceHref,
  workspaceMemberHref,
  workspaceMembersHref,
} from '../links';

interface RefModel {
  _links: Record<string, Link>;
  id: string;
}

export interface MemberModel {
  _links: Record<string, Link>;
  id: string;
  workspace: RefModel;
  user: RefModel;
  role: string;
  createdAt: string;
  updatedAt: string;
}

export function memberModel(userId: string, member: Member): MemberModel {
  const workspaceId = member.workspaceId();
  const memberUserId = member.description().user.id();
  return {
    _links: {
      self: link(workspaceMemberHref(userId, workspaceId, member.identity())),
      collection: link(workspaceMembersHref(userId, workspaceId)),
      workspace: link(workspaceHref(userId, workspaceId)),
      user: link(userHref(memberUserId)),
    },
    id: member.identity(),
    workspace: {
      _links: { self: link(workspaceHref(userId, workspaceId)) },
      id: workspaceId,
    },
    user: {
      _links: { self: link(userHref(memberUserId)) },
      id: memberUserId,
    },
    role: member.description().role,
    createdAt: member.createdAt(),
    updatedAt: member.updatedAt(),
  };
}

File: libs/server-java/api/src/main/java/reengineering/ddd/evidence/api/representation/MemberModel.java
package reengineering.ddd.evidence.api.representation;

import com.fasterxml.jackson.annotation.JsonProperty;
import jakarta.ws.rs.core.UriInfo;
import java.time.Instant;
import reengineering.ddd.evidence.api.ApiTemplates;
import reengineering.ddd.evidence.domain.description.MemberDescription;
import reengineering.ddd.evidence.domain.model.Member;

public final class MemberModel extends EvidenceModel<MemberModel> {
  @JsonProperty private final String id;
  @JsonProperty private final RefModel workspace;
  @JsonProperty private final RefModel user;
  @JsonProperty private final String role;
  @JsonProperty private final Instant createdAt;
  @JsonProperty private final Instant updatedAt;

  public MemberModel(Member member, UriInfo uriInfo) {
    MemberDescription description = member.getDescription();
    String workspaceId = description.workspace().id();
    String userId = description.user().id();
    id = member.getIdentity();
    workspace = RefModel.workspace(workspaceId, uriInfo);
    user = RefModel.user(userId, uriInfo);
    role = description.role();
    createdAt = description.createdAt();
    updatedAt = description.updatedAt();
    addSelf(ApiTemplates.workspaceMember(uriInfo, workspaceId, id));
    addRelation(ApiTemplates.workspaceMembers(uriInfo, workspaceId), "collection");
    addRelation(ApiTemplates.workspace(uriInfo, workspaceId), "workspace");
    addRelation(ApiTemplates.user(uriInfo, userId), "user");
  }

  public static final class RefModel extends EvidenceModel<RefModel> {
    @JsonProperty private final String id;

    private RefModel(String id) {
      this.id = id;
    }

    private static RefModel workspace(String id, UriInfo uriInfo) {
      RefModel model = new RefModel(id);
      model.addSelf(ApiTemplates.workspace(uriInfo, id));
      return model;
    }

    public static RefModel user(String id, UriInfo uriInfo) {
      RefModel model = new RefModel(id);
      model.addSelf(ApiTemplates.user(uriInfo, id));
      return model;
    }
  }
}
